website/addons/s3/views.py

```diff
@@ -0,0 +1,189 @@
+            oauth_key=access_key,
+            oauth_secret=secret_key,
+            provider_id=user_info.id,
+            display_name=user_info.display_name,
+        )
+        account.save()
+    except KeyExistsException:
+        # ... or get the old one
+        account = ExternalAccount.find_one(
+            Q('oauth_key', 'eq', access_key) &
+            Q('oauth_secret', 'eq', secret_key)
+        )
+    assert account is not None
+
+    if account not in auth.user.external_accounts:
+        auth.user.external_accounts.append(account)
+
+    # Ensure S3 is enabled.
+    auth.user.get_or_add_addon('s3', auth=auth)
+    auth.user.save()
+
+    return {}
+
+
+@must_have_permission('write')
+@must_have_addon(SHORT_NAME, 'user')
+@must_have_addon(SHORT_NAME, 'node')
+@must_be_addon_authorizer(SHORT_NAME)
+@must_not_be_registration
+def s3_set_config(node, auth, user_addon, node_addon, **kwargs):
+    """Saves selected bucket to node settings."""
+    # Fail if user settings not authorized
+    if not user_addon.has_auth:
+        raise HTTPError(httplib.BAD_REQUEST)
+
+    # If authorized, only owner can change settings
+    if node_addon.has_auth and node_addon.user_settings.owner != auth.user:
+        raise HTTPError(httplib.BAD_REQUEST)
+
+    # Claiming the node settings
+    if not node_addon.user_settings:
+        node_addon.user_settings = user_addon
+
+    bucket = request.json.get('s3_bucket', '')
+
+    if not utils.bucket_exists(node_addon.external_account.oauth_key, node_addon.external_account.oauth_secret, bucket):
+        error_message = ('We are having trouble connecting to that bucket. '
+                         'Try a different one.')
+        return {'message': error_message}, httplib.BAD_REQUEST
+
+    if bucket != node_addon.bucket:
+
+        # Update node settings and log
+        node_addon.set_folder(bucket, auth)
+
+    return S3Serializer().serialize_settings(node_addon, auth.user)
+
+@must_be_addon_authorizer(SHORT_NAME)
+@must_have_addon('s3', 'node')
+@must_have_permission('write')
+def create_bucket(auth, node_addon, **kwargs):
+    bucket_name = request.json.get('bucket_name', '')
+    bucket_location = request.json.get('bucket_location', '')
+
+    if not utils.validate_bucket_name(bucket_name):
+        return {
+            'message': 'That bucket name is not valid.',
+            'title': 'Invalid bucket name',
+        }, httplib.NOT_ACCEPTABLE
+
+    # Get location and verify it is valid
+    if not utils.validate_bucket_location(bucket_location):
+        return {
+            'message': 'That bucket location is not valid.',
+            'title': 'Invalid bucket location',
+        }, httplib.NOT_ACCEPTABLE
+
+    try:
+        utils.create_bucket(node_addon, bucket_name, bucket_location)
+    except exception.S3ResponseError as e:
+        return {
+            'message': e.message,
+            'title': 'Problem connecting to S3',
+        }, httplib.NOT_ACCEPTABLE
```

Comment:
These should all be 400s. Unless its a pain to change the error handling code client side.